File: 2014년문제답안/Problem02/Graduates/sequential.c
#include<stdio.h>
#include<stdlib.h>
#include<stddef.h>
#include<string.h>
#include<math.h>
#include<sys/time.h>




typedef struct Pos{
	float x,y,z;
}Pos;

double potential(Pos *r, int np){
	int i,j;
	double potent = 0;
	for(j=0;j<np;j++){
		for(i=j+1;i<np;i++){
			float x= r[i].x-r[j].x;
			float y= r[i].y-r[j].y;
			float z= r[i].z-r[j].z;
			float dist = sqrtf(x*x+y*y+z*z);
			if(dist > 0.) potent += 1./dist;
		}
	}
	return potent;
}

int getrandomnp(int istep, int niter){
	if(istep%3 ==0) return (int)(50000*(rand()/(RAND_MAX+1.)));
	else  return (int)(500*(rand()/(RAND_MAX+1.)));
}

struct timeval tv;
float gettime(){
	static int startflag = 1;
	static double tsecs0, tsecs1;
	if(startflag) {
		(void ) gettimeofday(&tv, NULL);
		tsecs0 = tv.tv_sec + tv.tv_usec*1.0E-6;
		startflag = 0;
	}
	(void) gettimeofday(&tv, NULL);
	tsecs1 = tv.tv_sec + tv.tv_usec*1.0e-6;
	return (float) (tsecs1 - tsecs0);
}



int main(int argc, char **argv){
	int i, j;
	int np,niter,maxnp=5000000;
	Pos *r;
	double totpotent=0;
	unsigned int iseed = 100;

	niter = 50;
	r = (Pos*)malloc(sizeof(Pos)*maxnp);
	srand( iseed);

	float time1, time2;
	time1 = gettime();



	for(i=0;i<niter;i++){
		np =  getrandomnp(i,niter);
		for(j=0;j<np;j++){
			r[j].x = 2.*(rand()/(RAND_MAX+1.))-1.;
			r[j].y = 2.*(rand()/(RAND_MAX+1.))-1.;
			r[j].z = 2.*(rand()/(RAND_MAX+1.))-1.;
		}
		totpotent += potential(r, np);
	}
	time2 = gettime();
	printf("Total potential is %20.10g in wallclock time = %g second\n",totpotent, (time2-time1));
}

File: 2014년문제답안/Problem02/Undergraduates/sequential.f
c234567
      function potential(x,y,z, np) 
      real*8 potential
      real*4 x(np),y(np),z(np),dist,tmpx,tmpy,tmpz
      potential = 0
      do j = 1, np
      do i = j+1, np
         tmpx = x(i)-x(j)
         tmpy = y(i)-y(j)
         tmpz = z(i)-z(j)
         dist = sqrt(tmpx*tmpx + tmpy*tmpy + tmpz*tmpz)
		 if(dist .gt.0) potential = potential + 1./dist
      enddo
      enddo
      return 
      end

      function getrandomnp(istep, niter)
      integer getrandomnp,istep,niter
      if(istep .lt. niter/10) then
         getrandomnp = (3000+5000*rand())
      else
         getrandomnp = (100*rand())
      endif
      return
      end


      program main
	  parameter(maxnp= 5000000)
      integer i,j,np,niter
      real*8 totpotent,potential
      integer iseed
      real x(maxnp),y(maxnp),z(maxnp)
      real timearray(2),time1,time2,walltime
	  integer getrandomnp
	  external getrandomnp



	  totpotent = 0
	  iseed = 100
      niter = 4000
      call srand(iseed)


      call etime(timearray, time1)
      do i = 1, niter
         np = getrandomnp(i,niter)
         do j = 1, np
            x(j) = 2.*rand()-1.
            y(j) = 2.*rand()-1.
            z(j) = 2.*rand()-1.
         enddo
         totpotent = totpotent + potential(x,y,z,np)
      enddo
      call etime(timearray, time2)
      walltime = time2-time1
      print *,'total potential =',totpotent,' and wallclock time=',
     &          walltime
      stop
      end
